lib/reader_writer.toit

```diff
@@ -80,27 +76,25 @@ monitor ReaderWriterHelper_:
     if writer_closed_: throw "CLOSED"
     result := to - from
     while from != to:
-      await: fullness_ != buffer_.size or outgoing_ == null or reader_closed_
+      await: fullness_ != buffer_.size or reader_closed_
       if reader_closed_: throw "CLOSED"
       space := buffer_.size - fullness_
-      if space == 0:
-        assert: outgoing_ == null
-        outgoing_ = buffer_
-        buffer_ = ByteArray buffer_size_
-        fullness_ = 0
-        space = buffer_.size
       remaining := to - from
       chunk_size := min space remaining
+      // Put as much as possible into the buffer.
+      // If we fill it up entirely, then the 'await' above will make us
+      // wait until a reader empties the buffer.
       buffer_.replace fullness_ data from from + chunk_size
       from += chunk_size
       fullness_ += chunk_size
     return result
 
   read:
-    await: outgoing_ or writer_closed_
-    if outgoing_:
-      result := outgoing_
-      outgoing_ = null
+    await: fullness_ != 0 or writer_closed_
+    result := ?
+    if fullness_ != 0:
+      result = buffer_[..fullness_].copy
```

Comment:
I think copy takes from and to, so you are not creating a slice object here for no reason.

Reply:
done.